Error: Vehicle has no corresponding soldier when trying to take control over Meduna.

#### Bug description
2026-08-02T13:51:19.3728193Z [INFO] stracciatella::logger: Logging to file "C:\\Users\\Bezi\\AppData\\Local\\Temp\\ja2.log"
2026-08-02T13:51:19.4025485Z [WARN] stracciatella::mods: Error loading mod directory: Mod directory Path("D:\\Gry\\Jagged Alliance 2\\JA2 Stracciatella\\mods\\mod-wildfire-maps-0.5.2") does conform to mod id format: ^[a-z0-9\-]+$
2026-08-02T13:51:19.8578547Z [INFO] stracciatella::vfs: VFS layer 1: SlfFs { source: DirFsFile { "tilesets.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8588919Z [INFO] stracciatella::vfs: VFS layer 2: SlfFs { source: DirFsFile { "speech.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8593687Z [INFO] stracciatella::vfs: VFS layer 3: SlfFs { source: DirFsFile { "sounds.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.85977Z [INFO] stracciatella::vfs: VFS layer 4: SlfFs { source: DirFsFile { "radarmaps.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.860171Z [INFO] stracciatella::vfs: VFS layer 5: SlfFs { source: DirFsFile { "polish.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8607239Z [INFO] stracciatella::vfs: VFS layer 6: SlfFs { source: DirFsFile { "npcdata.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8611451Z [INFO] stracciatella::vfs: VFS layer 7: SlfFs { source: DirFsFile { "npc_speech.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8614698Z [INFO] stracciatella::vfs: VFS layer 8: SlfFs { source: DirFsFile { "music.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8622961Z [INFO] stracciatella::vfs: VFS layer 9: SlfFs { source: DirFsFile { "mercedt.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8629023Z [INFO] stracciatella::vfs: VFS layer 10: SlfFs { source: DirFsFile { "maps.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8632687Z [INFO] stracciatella::vfs: VFS layer 11: SlfFs { source: DirFsFile { "loadscreens.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8636422Z [INFO] stracciatella::vfs: VFS layer 12: SlfFs { source: DirFsFile { "laptop.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8640908Z [INFO] stracciatella::vfs: VFS layer 13: SlfFs { source: DirFsFile { "intro.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8645485Z [INFO] stracciatella::vfs: VFS layer 14: SlfFs { source: DirFsFile { "interface.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8655965Z [INFO] stracciatella::vfs: VFS layer 15: SlfFs { source: DirFsFile { "fonts.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8661315Z [INFO] stracciatella::vfs: VFS layer 16: SlfFs { source: DirFsFile { "faces.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8665Z [INFO] stracciatella::vfs: VFS layer 17: SlfFs { source: DirFsFile { "data.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8668165Z [INFO] stracciatella::vfs: VFS layer 18: SlfFs { source: DirFsFile { "cursors.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.867132Z [INFO] stracciatella::vfs: VFS layer 19: SlfFs { source: DirFsFile { "binarydata.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8674404Z [INFO] stracciatella::vfs: VFS layer 20: SlfFs { source: DirFsFile { "bigitems.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8677904Z [INFO] stracciatella::vfs: VFS layer 21: SlfFs { source: DirFsFile { "battlesnds.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8684605Z [INFO] stracciatella::vfs: VFS layer 22: SlfFs { source: DirFsFile { "anims.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8687715Z [INFO] stracciatella::vfs: VFS layer 23: SlfFs { source: DirFsFile { "ambient.slf" in "D:\\Gry\\Jagged Alliance 2\\data" } }
2026-08-02T13:51:19.8693262Z [INFO] stracciatella::vfs: VFS layer 24: DirFs { "D:\\Gry\\Jagged Alliance 2\\data" }
2026-08-02T13:51:19.8697033Z [INFO] stracciatella::vfs: VFS layer 25: DirFs { "D:\\Gry\\Jagged Alliance 2\\JA2 Stracciatella\\externalized" }
2026-08-02T13:51:19.8701668Z [INFO] stracciatella::vfs: VFS layer 26: DirFs { "D:\\Gry\\Jagged Alliance 2\\JA2 Stracciatella\\mods\\imp-quiz-honest-answers\\data" }
2026-08-02T13:51:19.8706495Z [INFO] externalized\DefaultContentManager.cc: JA2 Home Dir:                  'C:\Users\Bezi\AppData\Roaming\JA2'
2026-08-02T13:51:19.8711304Z [INFO] externalized\DefaultContentManager.cc: Root game resources directory: 'D:\Gry\Jagged Alliance 2'
2026-08-02T13:51:19.8714923Z [INFO] externalized\DefaultContentManager.cc: Extra data directory:          'D:\Gry\Jagged Alliance 2\JA2 Stracciatella'
2026-08-02T13:51:19.8718031Z [INFO] externalized\DefaultContentManager.cc: Saved games directory:         'D:\Gry\Jagged Alliance 2\SavedGames'
2026-08-02T13:51:19.8721115Z [INFO] externalized\DefaultContentManager.cc: Temporary directory:           'C:\Users\Bezi\AppData\Local\Temp\ja2-stracciatella-K2MgU4eY'
2026-08-02T13:51:19.872424Z [INFO] externalized\ModPackContentManager.cc: Enabled mods                    'imp-quiz-honest-answers'
2026-08-02T13:51:20.4197533Z [INFO] game\GameLoop.cc: Version Label: Stracciatella v0.23.0-20260715+b71c5d3
2026-08-02T13:51:20.4205526Z [INFO] game\GameLoop.cc: Version #:     Build 04.12.02
2026-08-02T13:51:20.4382458Z [WARN] game\Tactical\Interface_Panels.cc: The team panel does not have enough space to display all soldiers in a squad. You should either increase the game resolution or decrease the squad size
2026-08-02T13:52:30.9690943Z [ERROR] game\Strategic\Strategic_AI.cc: Attempting to send reinforcements from a garrison that doesn't have any! -- KM:0 (with prior saved game and Debug Log)
2026-08-02T13:53:11.7770512Z [ERROR] game\Strategic\Strategic_AI.cc: Attempting to send reinforcements from a garrison that doesn't have any! -- KM:0 (with prior saved game and Debug Log)
2026-08-02T13:54:16.5248618Z [ERROR] sgp\SGP.cc: Game has been terminated due to an unrecoverable error: Vehicle has no corresponding soldier
Creating an emergency savegame succeeded (Error.sav).
Please report this error with a description of the circumstances.  Do not forget to attach the savegame. (class std::runtime_error)

[Error.zip](https://github.com/user-attachments/files/30633003/Error.zip)

#### Steps to reproduce
1. Attack any sector near Meduna
2a. Win fight.
3a. Try to open map
4.a Crash

2b Don't fight just move to another secord.
3b. Crash

Game version JA2 from GoG. 
2026-08-02T14:01:28.3671255Z [INFO] game\GameLoop.cc: Version Label: Stracciatella v0.23.0-20260715+b71c5d3
2026-08-02T14:01:28.3674788Z [INFO] game\GameLoop.cc: Version #:     Build 04.12.02

--- a/src/game/Strategic/Map_Screen_Interface_Map.cc
+++ b/src/game/Strategic/Map_Screen_Interface_Map.cc
@@ -724,9 +724,7 @@ static INT32 ShowVehicles(const SGPSector& sSector, INT32 icon_pos)
 		if (v.sSector != sSector)                     continue;
 		if (v.fDestroyed)                             continue;
 		if (PlayerIDGroupInMotion(v.ubMovementGroup)) continue;
-
-		SOLDIERTYPE const& vs = GetSoldierStructureForVehicle(v);
-		if (vs.bTeam != OUR_TEAM) continue;
+		if (auto s = FindSoldierType(v); !s || s->bTeam != OUR_TEAM) continue;
 
 		DrawMapBoxIcon(guiCHARICONS, SMALL_WHITE_BOX, sSector, icon_pos++);
 	}

--- a/src/game/Tactical/Vehicles.cc
+++ b/src/game/Tactical/Vehicles.cc
@@ -709,14 +709,21 @@ INT8 RepairVehicle(VEHICLETYPE const& v, INT8 const bRepairPtsLeft, BOOLEAN* con
 }
 
 
-SOLDIERTYPE& GetSoldierStructureForVehicle(VEHICLETYPE const& v)
+SOLDIERTYPE * FindSoldierType(VEHICLETYPE const& v)
 {
 	FOR_EACH_SOLDIER(s)
 	{
 		if (!(s->uiStatusFlags & SOLDIER_VEHICLE)) continue;
 		if (s->bVehicleID != VEHICLE2ID(v))        continue;
-		return *s;
+		return s;
 	}
+	return nullptr;
+}
+
+
+SOLDIERTYPE& GetSoldierStructureForVehicle(VEHICLETYPE const& v)
+{
+	if (auto pSoldier = FindSoldierType(v); pSoldier) return *pSoldier;
 	throw std::logic_error("Vehicle has no corresponding soldier");
 }
 

--- a/src/game/Tactical/Vehicles.h
+++ b/src/game/Tactical/Vehicles.h
@@ -102,6 +102,9 @@ BOOLEAN ExitVehicle( SOLDIERTYPE *pSoldier );
 void VehicleTakeDamage(UINT8 ubID, UINT8 ubReason, INT16 sDamage, INT16 sGridNo, SOLDIERTYPE* att);
 
 // the soldiertype containing this tactical incarnation of this vehicle
+// This version returns nullptr if there is no corresponding soldier.
+SOLDIERTYPE * FindSoldierType(VEHICLETYPE const& v);
+// This version throws an exception if there is no corresponding soldier.
 SOLDIERTYPE& GetSoldierStructureForVehicle(VEHICLETYPE const&);
 
 // does it need fixing?

